MudDataGrid

tjscience opened the pull request:
Completed the first phase of the MudDataGrid component. 

1. Created the parent component and all required children. 
2. Ensured that there is backward compatability with the MudTable and most of it's use cases as this component is poised to replace the MudTable.
3. Added documentation under the "Data Grid" section (note that I did not recreate all of the examples that exist currently in MudTable as of yet. This will be coming later).
4. Added test cases to push coverage above 86% (more to come as well).

Phase 2:

1. Sort multiple columns at once.
2. DataGridEditMode.Cell mode to allow excel like editing by cell.
3. Built in aggregation for footers.
4. Column resizing.
5. Column chooser.
6. More built in filter data types.

relates to #1626, closes #506

tjscience committed:
Added MudDataGrid component and all associated components/logic.

tjscience committed:
Fixed merge conflicts

tjscience committed:
Fixed some other merging issues.

tjscience committed:
Merge branch 'MudBlazor:dev' into table_filtering

tjscience committed:
Completed MudDataGrid phase 1.

tjscience committed:
Completed MudDataGrid phase 1 - merge conflic resolution.

tjscience committed:
Completed MudDataGrid component phase 1 - cleaned up MudTable examples.

tjscience committed:
Commented out the DataGridEventCallbacksTest test because it was causing the test run github action to fail.

tjscience committed:
Fixed 'Delegate to an instance method cannot have null 'this'.' issue.

JonBunator reviewed (commented):
Looks pretty good already :)
Much simpler code than the current table. I like the inlining of `<tr>,<td>...` This probably leads to a performance improvement.
I think we should merge this early in an experimental section. Then improve from there until it offers all the functionality of the current table

tjscience committed:
Merge branch 'MudBlazor:dev' into table_filtering

tjscience committed:
Removed unnecessary item group in MudBlazor.Docs project file.
Removed unnecessary item group from MudBlazor project file.
Removed workaround for z-inde of filter popovers.
Changed CompileSortBy in HeaderCell to execute with the SortBy getter.

tjscience committed:
Merge branch 'table_filtering' of https://github.com/tjscience/MudBlazor into table_filtering

Garderoben commented on src/MudBlazor.Docs/Pages/Components/DataGrid/DataGridPage.razor:
```diff
@@ -0,0 +1,75 @@
+                </Description>
+            </SectionHeader>
+            <SectionContent DarkenBackground="true" FullWidth="true">
+                <DataGridBasicExample />
```
This example is great but looks like the table even when sortable and filterable is not true this button still shows.
![image](https://user-images.githubusercontent.com/10367109/139533224-d4f8345c-bc20-4049-849f-d0147617e746.png)

tjscience replied:
This is actually by design but I guess it doesn't make much sense right now because the other options are not yet available. This will eventually include other options like Hide, Show, etc.

Garderoben commented on src/MudBlazor.Docs/Pages/Components/DataGrid/DataGridPage.razor:
```diff
@@ -0,0 +1,75 @@
+                </Description>
+            </SectionHeader>
+            <SectionContent DarkenBackground="true" FullWidth="true">
+                <DataGridAdvancedExample />
```
Please set the default rows per page to 4 in this example. Possible even remove the switch

tjscience replied:
Is this even possible with the current implementation of the data pager (MudDataGridPager is a replication of MudTablePager)? I have tried to set the PageSizeOptions and only include 4 there but the pager still uses 10 as the default. Maybe I am missing something there?

Garderoben reviewed (changes requested):
Apart from my minor comments, we should maybe think about `PagerContent `and `ToolBarContent` `FooterContent`, `HeaderContent `etc... can the datagrid only have ChildContent as renderfragment? Then we could create a standalone component like MudDataGridContent and have Class/Style Parameter. Then we dont need stupid things like `HorizontalAlignment` on the Pager itself

Garderoben commented on src/MudBlazor/Components/DataGrid/MudDataGrid.razor:
```diff
@@ -0,0 +1,170 @@
+                                            Id="@f.Id"></Filter>
+                                    }
+                                </MudGrid>
+                                <MudButton Class="mt-2" StartIcon="@Icons.Material.Filled.Add" OnClick="@AddFilter" Color="@Color.Info">Add Filter</MudButton>
```
Should be `Color.Primary`, and should probably use a Popover to get the flipping behavior and prevent any overflow issues.

tjscience replied:
I can change the color, but I had tons of problems when I first tried to do this with Popover. Any help in getting a Popover implementation to work would be appreciated.

tjscience committed:
Hid column options in data grid header cells when the cell is not sortable or filterable.
Changed the Add Filter button to Color.Primary.

cocytus commented on src/MudBlazor/Components/DataGrid/MudDataGrid.razor.cs:
```diff
@@ -0,0 +1,797 @@
+        {
+            StartedCommittingItemChangesEvent?.Invoke(item);
+            // Here, we need to validate at the cellular level...
+            await StartedCommittingItemChanges.InvokeAsync(item);
```
Would it be nice to wrap `await StartedCommittingItemChanges.InvokeAsync(item);` in try catch here, like:

```csharp
        internal async Task CommitItemChangesAsync(T item)
        {
            StartedCommittingItemChangesEvent?.Invoke(item);
            try
            {
                // Here, we need to validate at the cellular level...
                await StartedCommittingItemChanges.InvokeAsync(item);
                ClearEditingItem();
            }
            catch { /* Commit cancelled */ }
        }
```

Use case is code that is wants to save the updated row, but the save fails for any reason. This would allow the user to make changes to the row and then attempt to commit again.

segfault- commented:
i've been using @tjscience branch in my application.  I did add some minor things to support server side filtering and sorting.  Also, took me a minute to discover using a <CellTemplate> inside a column.  An additional example for that use case would be helpful.  I'll work on getting a PR to @tjscience 's fork so he can review the changes i made.

henon commented:
Like @Garderoben already said, we anticipate lots of feature requests for this new table in the future. Table is in our experience a very "hot" component with hundreds of different use-cases. In this light and to avoid lots of issues and PRs in the future, we think it essential to make almost everything a RenderFragment component that can be overridden by the user. Also, we'll have to think of natural extension hooks in the form of Func and EventCallback parameters.

henon commented:
It would lower overall coverage by -1.47% right now, so that is something that needs to be addressed. 

@Garderoben any changes that you deem necessary before merging?

Garderoben commented:
> So, what is the plan here? Are you looking for all of these changes in this pull request before you merge, or can these be done separately? I have already been working on other features and was hoping the data grid would be merged and available in an experimental fashion as was discussed before. This would also make it easier to PR future changes.

Sorry i been quite busy with some other tasks, one of them being .NET 6 release, i will look over everything and answer you later today. But as @henon already said, the coverage needs to be addressed.

tjscience commented:
Okay. I was originally told it had to stay above 86%, which my changes adhere to, but I will see if I can get the test coverage up a bit.

tjscience committed:
Couple of small fixes.
Added many more tests.

Garderoben commented:
> Okay. I was originally told it had to stay above 86%, which my changes adhere to, but I will see if I can get the test coverage up a bit.

That was probably the case then, but the higher it gets the higher the bar we set hehe.

I think we can merge this, but just to be clear, that we will slam it under a experimental section. And we might change some stuff that we seen dident turn  out that good with current table.

When you have done your changes maybe we can set a date that we merge it, as i said before i dont wanna loose you or your work, and we will be very strict around this new table since we don't want it to end up like current one.

tjscience commented:
> > Okay. I was originally told it had to stay above 86%, which my changes adhere to, but I will see if I can get the test coverage up a bit.
> 
> That was probably the case then, but the higher it gets the higher the bar we set hehe.
> 
> I think we can merge this, but just to be clear, that we will slam it under a experimental section. And we might change some stuff that we seen dident turn out that good with current table.
> 
> When you have done your changes maybe we can set a date that we merge it, as i said before i dont wanna loose you or your work, and we will be very strict around this new table since we don't want it to end up like current one.

I am done with my phase I changes and I also just checked in a couple of fixes as well as many more tests which should boost coverage to a better percentage. I am working on phase 2 solely from this point on.

Garderoben commented:
> > > Okay. I was originally told it had to stay above 86%, which my changes adhere to, but I will see if I can get the test coverage up a bit.
> > 
> > 
> > That was probably the case then, but the higher it gets the higher the bar we set hehe.
> > I think we can merge this, but just to be clear, that we will slam it under a experimental section. And we might change some stuff that we seen dident turn out that good with current table.
> > When you have done your changes maybe we can set a date that we merge it, as i said before i dont wanna loose you or your work, and we will be very strict around this new table since we don't want it to end up like current one.
> 
> I am done with my phase I changes and I also just checked in a couple of fixes as well as many more tests which should boost coverage to a better percentage. I am working on phase 2 solely from this point on.

Codecov still complains it seems.

henon commented:
> Codecov still complains it seems.

It complains because we set the rules to 100% coverage for the patch with 10% leniency.  The patch has only 86.5% though.

tjscience committed:
Added more tests...

tjscience committed:
Fixed an issue with SortBy.

tjscience committed:
Merge branch 'MudBlazor:dev' into table_filtering

Garderoben commented:
@tjscience is this ready to be merged from your side? if so i will prepare the docs and talk to the team first, then merge.

tjscience commented:
@Garderoben, yes.

Eagle3386 commented:
As I'm eagerly waiting for using MudDataGrid, may I ask for a rough estimate regarding 6.0.3's release?

Garderoben commented:
> As I'm eagerly waiting for using MudDataGrid, may I ask for a rough estimate regarding 6.0.3's release?

This year

Eagle3386 commented:
... and you delivered, just right on schedule! 😅
Thanks so much for this, @Garderoben! 🍻🤜🏻

Garderoben commented:
> ... and you delivered, just right on schedule! 😅 Thanks so much for this, @Garderoben! 🍻🤜🏻

haha i rarely do, things i say will be done the next two days is always delayed for some reason :( but i really really wanted to get this out, happy you even remembered <3

Eagle3386 commented:
Of course - looking _so_ much forward to `MudDataGrid`! 😍
And I'm even more thankful that you took on building a Razor-based framework that tries its best to avoid JS whenever possible while simultaneously aiming for "C# + HTML (with a bit of CSS sometimes 😉)" - it's a dream coming true for me. Seriously! ❤️‍🔥

neozhu commented:
I saw the sample QuickFilter does not support the server-side function, right?

segfault- commented:
> I saw the sample QuickFilter does not support the server-side function, right?

correct - i was able to implement server side by modifying gridstate to track sort field and filter definitions and added datagrid filter ui apply button that tiggers InvokeServerLoadFunc.

engsky commented:
It is better to add new row by Tab from last column and then after adding new row the focus should set within the first column of the added row.
